File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scbool.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

const (
	ScBoolLength = 1
	ScBoolFalse  = 0x00
	ScBoolTrue   = 0xff
)

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func BoolDecode(dec *WasmDecoder) bool {
	return dec.Byte() != ScBoolFalse
}

func BoolEncode(enc *WasmEncoder, value bool) {
	if value {
		enc.Byte(ScBoolTrue)
		return
	}
	enc.Byte(ScBoolFalse)
}

func BoolFromBytes(buf []byte) bool {
	if buf == nil {
		return false
	}
	if len(buf) != ScBoolLength {
		panic("invalid Bool length")
	}
	if buf[0] == ScBoolFalse {
		return false
	}
	if buf[0] != ScBoolTrue {
		panic("invalid Bool value")
	}
	return true
}

func BoolToBytes(value bool) []byte {
	if value {
		return []byte{ScBoolTrue}
	}
	return []byte{ScBoolFalse}
}

func BoolToString(value bool) string {
	if value {
		return "1"
	}
	return "0"
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableBool struct {
	proxy Proxy
}

func NewScImmutableBool(proxy Proxy) ScImmutableBool {
	return ScImmutableBool{proxy: proxy}
}

func (o ScImmutableBool) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableBool) String() string {
	return BoolToString(o.Value())
}

func (o ScImmutableBool) Value() bool {
	return BoolFromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableBool struct {
	ScImmutableBool
}

func NewScMutableBool(proxy Proxy) ScMutableBool {
	return ScMutableBool{ScImmutableBool{proxy: proxy}}
}

func (o ScMutableBool) Delete() {
	o.proxy.Delete()
}

func (o ScMutableBool) SetValue(value bool) {
	o.proxy.Set(BoolToBytes(value))
}

File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scbool.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import {WasmDecoder, WasmEncoder} from "./codec";
import {Proxy} from "./proxy";

export const ScBoolLength = 1;
export const ScBoolFalse = 0x00
export const ScBoolTrue = 0x01

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function boolDecode(dec: WasmDecoder): bool {
    return dec.byte() != ScBoolFalse;
}

export function boolEncode(enc: WasmEncoder, value: bool): void {
    enc.byte((value ? ScBoolTrue : ScBoolFalse) as u8);
}

export function boolFromBytes(buf: Uint8Array): bool {
    if (buf.length == 0) {
        return false;
    }
    if (buf.length != ScBoolLength) {
        panic("invalid Bool length");
    }
    if (buf[0] == ScBoolFalse) {
        return false;
    }
    if (buf[0] != ScBoolTrue) {
        panic("invalid Bool value");
    }
    return true;
}

export function boolToBytes(value: bool): Uint8Array {
    const buf = new Uint8Array(ScBoolLength);
    buf[0] = (value ? ScBoolTrue : ScBoolFalse) as u8;
    return buf;
}

export function boolFromString(value: string): bool {
    if (value == "0") {
        return false;
    }
    if (value == "1") {
        return true;
    }
    panic("invalid bool string");
    return false;
}

export function boolToString(value: bool): string {
    return value ? "1" : "0";
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableBool {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return boolToString(this.value());
    }

    value(): bool {
        return boolFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableBool extends ScImmutableBool {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: bool): void {
        this.proxy.set(boolToBytes(value));
    }
}
